Makefile

```diff
@@ -309,17 +309,72 @@ run: install
 
-install: manifests kustomize ## Install CRDs into the K8s cluster specified in ~/.kube/config.
+# ------------------------------------------------------------------------------
+# Gateway API
+# ------------------------------------------------------------------------------
+
+GATEWAY_API_PACKAGE ?= sigs.k8s.io/gateway-api
+# TODO: Below hardcoded ref is a workaround for the fact that we're using an untagged version
+#       of sigs.k8s.io/gateway-api in go.mod - that occurred after v0.4.0 (which was tagged on master)
+#       but which contains a breaking change w.r.t to the file structure in said repo - and the
+#       fact that kustomize accepts only branch names, tags, or full commit hashes, i.e. short
+#       hashes or go pseudo versions are not supported [1].
+#       Please also note that kustomize fails silently when provided with an unsupported ref
+#       and downloads the manifests from the main branch.
+#
+#       [1]: https://github.com/kubernetes-sigs/kustomize/blob/master/examples/remoteBuild.md#remote-directories
+#
+#       This causes a problem where we cannot use go pseudo version from go.mod i.e.
+#       v0.4.1-0.20220306235253-71fee1c2808f and where we cannot update to a newer version
+#       sigs.k8s.io/gateway-api because v0.5.0 hasn't been released yet and v0.4.x versions
+#       do not contain the change in file structure that some of the code in this repo already
+#       relies on.
+#
+#       In order to avoid unnecessary work we're just hardcoding the full SHA that
+#       corresponds to what's in go.mod - v0.4.1-0.20220306235253-71fee1c2808f - until
+#       v0.5.0 is released which we can then use in go.mod and scrape via go list ...
+GATEWAY_API_VERSION ?= 71fee1c2808fa19a5f19d952d155fc072cf9324c
+# GATEWAY_API_VERSION ?= $(shell go list -m -f "{{.Version}}" $(GATEWAY_API_PACKAGE))
+GATEWAY_API_CRDS_LOCAL_PATH = $(shell go env GOPATH)/pkg/mod/$(GATEWAY_API_PACKAGE)@$(GATEWAY_API_VERSION)/config/crd
+GATEWAY_API_REPO ?= github.com/kubernetes-sigs/gateway-api
+GATEWAY_API_CRDS_URL = $(GATEWAY_API_REPO)/config/crd?ref=$(GATEWAY_API_VERSION)
+
+.PHONY: print-gateway-api-crds-url
+print-gateway-api-crds-url:
+	@echo $(GATEWAY_API_CRDS_URL)
+
+.PHONY: generate.gateway-api-crds-url
+generate.gateway-api-crds-url:
+	URL=$(shell $(MAKE) print-gateway-api-crds-url) \
+		INPUT=$(shell pwd)/test/internal/cmd/generate-gateway-api-crds-url/gateway_consts.tmpl \
+		OUTPUT=$(shell pwd)/test/consts/gateway.go \
+		go generate ./test/internal/cmd/generate-gateway-api-crds-url
+
+.PHONY: go-mod-download-gateway-api
+go-mod-download-gateway-api:
+	@go mod download $(GATEWAY_API_PACKAGE)
+
+.PHONY: install-gateway-api-crds
+install-gateway-api-crds: go-mod-download-gateway-api
+	$(KUSTOMIZE) build $(GATEWAY_API_CRDS_LOCAL_PATH) | kubectl apply -f -
+
+.PHONY: uninstall-gateway-api-crds
+uninstall-gateway-api-crds: go-mod-download-gateway-api
+	$(KUSTOMIZE) build $(GATEWAY_API_CRDS_LOCAL_PATH) | kubectl delete -f -
```

Comment:
I feel what you're saying but I don't have a solution for that.

We could separate the targets into multiple `Makefile`s and the use `include`. We could also create a script but I'm not sure what we'd gain from it other than less targets in `Makefile`. 🤔

Reply:
> We could also create a script but I'm not sure what we'd gain from it other than less targets in Makefile

As a matter of preference I think it's more ideal for `Makefiles` to only include higher level actions as much as feasible and to avoid a `Makefile` getting into a state where a lot of the file is basically just a sprawl of what is essentially "library code". Therefore the value proposition I see is easier maintenance through abstraction: code readers of the `Makefile` are shown the very specific actions they can take toplevel, but the details of how those actions work are behind a layer that is optional to investigate. It's really just an age old code composition problem compounded with the script-y nature of `Makefiles` which allows things to get pretty esoteric and hard to understand at a glance if you let them. ALL that said... I don't think what we have here right now is egregious, and if you're not strongly inspired to make any specific changes then I don't think it's worth spending a lot more time on right now. We can always come back around and refactor the `Makefile` in the future, given that it's non-user-facing.

Reply:
👍 

Generally I believe code structure is as important as functionality at times so you're making a very good point here. OTOH as you've said, let's leave this as is for now (hopefully not forever 😬 ) and perhaps there's going to be a better approach on the horizon next time we visit this piece of code.